corehq/apps/translations/app_translations.py

```diff
@@ -312,164 +368,197 @@ def expected_bulk_app_sheet_rows(app, exclude_module=None, exclude_form=None):
                     row_type="Form",
-                    sheet_name=form_string,
+                    sheet_name=form_sheet_name,
                     languages=[form.name.get(lang) for lang in app.langs],
-                    # leave all
                     media_image=[form.icon_by_language(lang) for lang in app.langs],
                     media_audio=[form.audio_by_language(lang) for lang in app.langs],
                     unique_id=form.unique_id
-                )
+                ))
 
-                # Add form to the first street
-                rows[MODULES_AND_FORMS_SHEET_NAME].append(first_sheet_row)
+                rows[form_sheet_name] = get_form_question_rows(app.langs, form)
 
-                if form.form_type == 'shadow_form':
-                    continue
+    return rows
 
-                # Populate form sheet
-                rows[form_string] = []
 
-                itext_items = OrderedDict()
-                try:
-                    nodes = xform.itext_node.findall("./{f}translation")
-                except XFormException:
-                    nodes = []
-
-                for translation_node in nodes:
-                    lang = translation_node.attrib['lang']
-                    for text_node in translation_node.findall("./{f}text"):
-                        text_id = text_node.attrib['id']
-                        itext_items[text_id] = itext_items.get(text_id, {})
-
-                        for value_node in text_node.findall("./{f}value"):
-                            value_form = value_node.attrib.get("form", "default")
-                            value = ''
-                            for part in ItextValue.from_node(value_node).parts:
-                                if isinstance(part, ItextOutput):
-                                    value += "<output value=\"" + part.ref + "\"/>"
-                                else:
-                                    value += mark_safe(force_text(part).replace('&', '&amp;').replace('<', '&lt;').replace('>', '&gt;'))
-                            itext_items[text_id][(lang, value_form)] = value
-
-                for text_id, values in six.iteritems(itext_items):
-                    row = [text_id]
-                    for value_form in ["default", "audio", "image", "video"]:
-                        # Get the fallback value for this form
-                        fallback = ""
-                        for lang in app.langs:
-                            fallback = values.get((lang, value_form), fallback)
-                            if fallback:
-                                break
-                        # Populate the row
-                        for lang in app.langs:
-                            row.append(values.get((lang, value_form), fallback))
-                    # Don't add empty rows:
-                    if any(row[1:]):
-                        rows[form_string].append(row)
+def get_module_rows(langs, module):
```

Comment:
much thanks for this breakdown of methods.
Would it be good to also move this to a separate file `translations/app_translations/utils.py` considering their specific use case and purpose.

Reply:
I definitely agree this file would benefit from being split up, but decided to wait on it until I implement the upload part of this feature. I can imagine a couple of different approaches - just split out utils, split by upload-related vs download-related, split by the multiple-language download vs the new single-language download, other options... - and I think working with the upload code will give me a better idea of what makes the most sense.